### DISCUSSION TITLE
Can a FileSystem extension set icons for files in the FileSystem viewer/browser?

### DISCUSSION THREAD

**Discussion by hippietrail** [NationalSecurityAgency/ghidra#6953] [Q&A] [answered]
Using built-in filesystems I see that certain filetypes have associated icons. Not that many, but text files get an icon for instance:  
<img width="210" alt="image" src="https://github.com/user-attachments/assets/07e3861d-a76f-4bd4-bde8-582a60066753">

In my FileSystems extensions I'm developing for retro systems I often have info on filetypes and it would be nice to be able to get the icons to reflect that. But I can't find how and suspect it may not be possible. I think the code for the fs browser/viewer just goes by the filename extension, but I don't want to add or change the extensions on the files in my filesystems.

Is there a way to do this I'm not seeing?

**Comment by sbrptdev2** [comment, marked as answer]
You should be able to alter the `GIcon` associated with the items, AFAIK. Whether a explicit API exists from Ghidra itself I don't know, but it's all Swing under the hood. 

A cursory look gives some hints:

https://ghidra.re/ghidra_docs/api/ghidra/formats/gfilesystem/GIconProvider.html#getIconForFile(ghidra.formats.gfilesystem.GFile,ghidra.util.task.TaskMonitor)

https://ghidra.re/ghidra_docs/api/docking/widgets/tree/GTree.html

The documentation for examples and how things are built is very sparse, but fgrep will be your best companion for the time being.

`fgrep --include '*.java' -rn getIconForFile .` (from a clone of Ghidra's source) 

From `./Ghidra/Features/Base/src/main/java/ghidra/plugins/fsbrowser/filehandlers/ImageFSBFileHandler.java` you can get a good idea of how you could approach the issue:

```
	@Override
	public List<DockingAction> createActions() {
		DockingAction action = new ActionBuilder(FSB_VIEW_AS_IMAGE, context.plugin().getName())
				.withContext(FSBActionContext.class)
				.enabledWhen(ac -> ac.notBusy() && ac.getFileFSRL() != null)
				.popupMenuIcon(FSBIcons.VIEW_AS_IMAGE)
				.popupMenuPath("View As", "Image")
				.popupMenuGroup("G")
				.onAction(ac -> {
					FSRL fsrl = ac.getFileFSRL();
					if (fsrl != null) {
						ac.getTree()
								.runTask(monitor -> doViewAsImage(fsrl, ac.getSourceComponent(),
									monitor));
					}
				})
				.build();
		action.getPopupMenuData().setParentMenuGroup("C");

		return List.of(action);
	}

	void doViewAsImage(FSRL fsrl, Component parent, TaskMonitor monitor) {

		try (RefdFile refdFile = context.fsService().getRefdFile(fsrl, monitor)) {

			Icon icon = GIconProvider.getIconForFile(refdFile.file, monitor);
			if (icon == null) {
				Msg.showError(this, parent, "Unable To View Image",
					"Unable to view " + fsrl.getName() + " as an image.");
				return;
			}
			Swing.runLater(() -> {
				JLabel label = new GIconLabel(icon);
				JOptionPane.showMessageDialog(null, label, "Image Viewer: " + fsrl.getName(),
					JOptionPane.INFORMATION_MESSAGE);
			});
		}
		catch (IOException | CancelledException e) {
			FSUtilities.displayException(this, parent, "Error Viewing Image File", e.getMessage(),
				e);
		}
	}
```

In this case a contextual menu is created for image types, using `COMMON_IMAGE_EXTENSIONS` (for the file-names). There is no intelligent file content matching involved. 

Hopefully this helps you get on the right direction.

Edit: you can do whatever it is you want on the object itself (so eventually, you can get to whatever Icon property is associated with it and alter it accordingly). Remember to cleanup/deal with anything you create on runtime if it is not automatically destroyed/released in context.

**Comment by dev747368** [comment]
> I think the code for the fs browser/viewer just goes by the filename extension, but I don't want to add or change the extensions on the files in my filesystems.

You are correct, the icons for files in the file system browser tree are mainly determined by the filename extension.  See the code in FSBComponentProvider.java, method renderFile(), about line 195'ish.

The icons for actions in a popup menu or a button on the toolbar will be determined by the info you put in the DockingAction.  You can add actions to the file system browser by creating your own FSBFileHandler derived class.  As an example, I would point to the TextFSBFileHandler class.

**Reply by hippietrail** [reply]
Thanks. I might look at a way to communicate file type info directly from the FileSystem later on.
